src/core/shared_memory_manager.cc

```diff
@@ -0,0 +1,244 @@
+    std::unique_ptr<SharedMemoryInfo> shm_info(new SharedMemoryInfo(
+        name, shm_key, offset, byte_size, shm_fd, tmp_addr));
+    shared_memory_map_.insert(std::make_pair(
+        name, std::unique_ptr<SharedMemoryInfo>(new SharedMemoryInfo(
+                  name, shm_key, offset, byte_size, shm_fd, tmp_addr))));
+  }
+  catch (std::exception& ex) {
+    return Status(
+        RequestStatusCode::INVALID_ARG,
+        "Unable to register shared memory region successfully.");
+  }
+
+  return Status::Success;
+}
+
+Status
+SharedMemoryManager::UnregisterSharedMemory(const std::string& name)
+{
+  LOG_VERBOSE(1) << "Unregister() '" << name << "'";
+
+  auto it = shared_memory_map_.find(name);
+  if (it != shared_memory_map_.end()) {
+    void* tmp_addr;
+    RETURN_IF_ERROR(get_shm_addr(
+        it->second->shm_fd_, it->second->offset_, it->second->byte_size_,
+        &tmp_addr));
+    RETURN_IF_ERROR(mmap_cleanup(tmp_addr, it->second->byte_size_));
+
+    // if no other region with same shm_key then close
+    bool last_one = true;
+    for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
+         ++itr) {
+      if (itr->second->shm_key_ == it->second->shm_key_)
+        last_one = false;
+    }
+    if (last_one) {
+      RETURN_IF_ERROR(shm_close(it->second->shm_fd_));
+    }
+
+    // remove region info from shared_memory_map_
+    shared_memory_map_.erase(it);
+  } else {
+    return Status(
+        RequestStatusCode::INVALID_ARG,
+        "Cannot unregister shared memory region that has not been registered.");
+  }
+
+  return Status::Success;
+}
+
+SharedMemoryManager::SharedMemoryManager(
+    const std::shared_ptr<ServerStatusManager>& status_manager)
+    : status_manager_(status_manager)
+{
+}
+
+SharedMemoryManager::~SharedMemoryManager()
+{
+  UnregisterAllSharedMemory();
+}
+
+Status
+SharedMemoryManager::Create(
+    const std::shared_ptr<ServerStatusManager>& status_manager,
+    std::unique_ptr<SharedMemoryManager>* shared_memory_manager)
+{
+  // Not setting the smart pointer directly to simplify clean up
+  std::unique_ptr<SharedMemoryManager> tmp_manager(
+      new SharedMemoryManager(status_manager));
+  *shared_memory_manager = std::move(tmp_manager);
+
+  return Status::Success;
+}
+
+Status
+SharedMemoryManager::RegisterUnregisterSharedMemory(
+    const std::string& name, const std::string& shm_key, const size_t offset,
+    const size_t byte_size, ActionType type)
+{
+  // Serialize all operations that write/read current shared memory regions
+  std::lock_guard<std::mutex> lock(register_mu_);
+  if (type == ActionType::UNREGISTER) {
+    UnregisterSharedMemory(name);
+  } else {
+    RegisterSharedMemory(name, shm_key, offset, byte_size);
+  }
+
+  return Status::Success;
+}
+
+Status
+SharedMemoryManager::UnregisterAllSharedMemory()
+{
+  for (const auto& shm_info : shared_memory_map_) {
+    Status unregister_status = UnregisterSharedMemory(shm_info.first);
```

Comment:
Should go through and unregister as many as possible and just collect the names of those that don't unregister and return a single error message listing all that didn't unregister... and just use the Message from one of them I guess.